### PR TITLE
 ✨ [RUMF-827] sanitize tags 

### PR DESCRIPTION
**Pull request by BenoitZugmeyer**
## Motivation

Help the user to understand why its provided values appear differently in the Datadog UI

## Changes

Sanitize tags and warn user of issues in its provided values.

## Testing

- [x] Local
- [ ] Staging
- [x] Unit
- [ ] End to end

---

I have gone over the [contributing](https://github.com/DataDog/browser-sdk/blob/main/CONTRIBUTING.md) documentation.

### CODE REVIEW

File: packages/core/src/domain/configuration/tags.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  }
+  if (version) {
+    tags.push(buildTag('version', version))
+  }
+
+  return tags
+}
+
+export function buildTag(name: string, value: string) {
+  value = value.toLowerCase()
```
[the remark this answers is not in the document]
Comment (bcaudan) [reply packages/core/src/domain/configuration/tags.ts:24]: nit: if the conversion is automatically done on the server side, we may not even need to do it

File: packages/core/src/domain/configuration/tags.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    tags.push(buildTag('service', service))
+  }
+  if (version) {
+    tags.push(buildTag('version', version))
+  }
+
+  return tags
+}
+
+export function buildTag(name: string, value: string) {
```
Comment (bcaudan) [inline packages/core/src/domain/configuration/tags.ts:23]:
```suggestion
export function buildTag(key: string, value: string) {
```
in order to use the same naming as in the doc

File: packages/core/src/domain/configuration/tags.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  value = value.toLowerCase()
+
+  // See https://docs.datadoghq.com/getting_started/tagging/#defining-tags for tags syntax.
+  const valueSizeLimit = TAG_SIZE_LIMIT - name.length - 1
+  if (value.length > TAG_SIZE_LIMIT) {
```
Comment (bcaudan) [inline packages/core/src/domain/configuration/tags.ts:28]: > Tags can be up to 200 characters 

should not it be: 
```suggestion
  if (name.length + value.length + 1 > TAG_SIZE_LIMIT) {
```

File: packages/core/src/domain/configuration/tags.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (bcaudan) [reply packages/core/src/domain/configuration/tags.ts:24]: it seems odd to not warn for this one, no?
anything specific about it?

[the remark this answers is not in the document]
Comment (BenoitZugmeyer) [reply packages/core/src/domain/configuration/tags.ts:24]: If we follow that thought, we could skip all sanitization here: it is also done on the server. I just tried, and a `foo:bar!éE:` tag is transformed to `foo:bar__e:` (strangely, the last colon is accepted, contrary to the documentation)

We could simplify a bit my PR and print a warning if a tag doesn't fit the expected format, ex: `"env value doesn't respect tag requirements and will be sanitized", and don't try to sanitize it ourselves. WDYT?

[the remark this answers is not in the document]
Comment (bcaudan) [reply packages/core/src/domain/configuration/tags.ts:24]: sorry, I removed my previous comment after re-reading the goal of the task 😅 

as you said I think we could simplify by:

- warn the user if his value will be changed
- not be specific about the change 

and we should probably at least avoid them to highjack our tags system by putting a value like `xxx&otherTag=yyy' if it is still possible.
No strong opinion about performing all the sanitization ourself or not.

File: packages/core/src/domain/configuration/tags.ts (hunk 3 above)
Comment (BenoitZugmeyer) [reply packages/core/src/domain/configuration/tags.ts:28]: Ah, good catch. I meant to say, `value.length > valueSizeLimit`

**Review by bcaudan — APPROVED** [review]

**Pull request merged**